### ISSUE TITLE
Sparse use of KBs suggests embedded database

### ISSUE THREAD

**Issue by hickst** [clulab/reach#482]
We could save resources by using an embedded SQL database for the KBs, wrapped with an object cache for speed.

**Comment by MihaiSurdeanu** [comment]
Now that I think about it, I think this would unfortunately not work for NER, which shares the same resources. To guarantee fast executions, all available strings must be loaded in memory, in a HashTrie...

Nevertheless, I think we should try to reduce the memory footprint of the KBs! Ideas?

**Comment by hickst** [comment]
The Trie could just as easily be built from the DB as from flat files. And the Trie is not used by  Grounding, which currently uses about 20 additional in-memory hash tables to hold the 4M+ entries and the 4-8M+ keys, most of which could be eliminated with the embedded DB + a cache.

**Comment by MihaiSurdeanu** [comment]
Good point about the additional hashes needed by grounding. How big are those?

**Comment by hickst** [comment]
A proof-of-concept, using an external MySQL database, has been implemented as part of the [kbquery](https://github.com/clulab/kbquery) project. This could be ported to Reach for use with grounding and/or NER.

**Comment by hickst** [comment]
Renaming this issue as the goals are expanding:
1) normalize data tables for efficiency
2) retest embedded database(s)
3) make Slick queries more efficient or replace them

**Comment by hickst** [comment]
5. Add functionality:
    a. additional Reach transforms, configurable per label type
    b. add `rlookup` method taking additional `label` argument, i.e. `rlookup(text, label)`
